File: 152/solution.py
class Solution(object):
    def maxProduct(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        ans = nums[0]
        tmax = 1
        tmin = 1
        for i in range(len(nums)):
            temp = tmax*nums[i]
            tmax = max(tmax*nums[i], tmin*nums[i], nums[i])
            tmin = min(temp, tmin*nums[i], nums[i])
            ans = max(ans, tmax, tmin)
        return ans

File: 152/solution.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int ans = nums[0];
        int tmax = 1;
        int tmin = 1;
        for(int i=0; i<nums.size(); ++i){
            int temp = tmax*nums[i];
            tmax = max(temp, max(tmin*nums[i], nums[i]));
            tmin = min(temp, min(tmin*nums[i], nums[i]));
            ans = max(tmax, ans);
        }
        return ans;
    }
};
